### DISCUSSION TITLE
Typescript and Typestates

### DISCUSSION THREAD

**Discussion by iacoware** [statelyai/xstate#2168] [Q&A] [answered]
Hi everyone, I'm trying to get Typestates to work within `actions` without any success. Can anyone confirm that I'm doing everything correctly and this scenario isn't supported?

XState: 4.19.0
Typescript: 4.2.3
TS mode: strict

```ts
import { assign, createMachine } from "xstate"

type BoringCtx = { alwaysThere: string; first?: number; second?: string }
type FirstCtx = BoringCtx & { first: number }
type SecondCtx = BoringCtx & FirstCtx & { second: string }

type BoringTypestate = { value: "first"; context: FirstCtx } | { value: "second"; context: SecondCtx }

type GoFirst = { type: "GO_FIRST"; first: string }
type GoSecond = { type: "GO_SECOND"; second: string }
type BoringEvent = GoFirst | GoSecond

export default createMachine<BoringCtx, BoringEvent, BoringTypestate>(
    {
        id: "boring",
        initial: "first" as const,
        context: {
            alwaysThere: "I'm here",
        },
        states: {
            first: {
                on: {
                    GO_SECOND: {
                        target: "second" as const,
                        actions: ["setSecond"],
                    },
                },
            },
            second: {},
        },
    },
    {
        actions: {
            setSecond: assign<BoringCtx, BoringEvent>({
                second: (ctx, ev) => (ctx as SecondCtx).first.toString() + "second",
            }),
        },
    },
)
```

Ideally inside the `setSecond` action I shouldn't be forced to do a downcast to `SecondCtx`. What I'd like to do instead:

```ts
    actions: {
        setSecond: assign<SecondCtx, GoSecond>({
            second: (ctx, ev) => ctx.first.toString() + "second",
        }),
    }
```  

Is this possible right now? If not, is it going to be supported in V5?

**Comment by Andarist** [comment, marked as answer]
> Is this possible right now?

No, typestates only affect matches calls

> If not, is it going to be supported in V5?

I certainly hope so! We have lots of great ideas how TS support can be improved and stroct - based on the accompanying typegen tool/IDE extension. The future is great - we just need to wait a little bit for it 😅

**Reply by iacoware** [reply]
Ok, got it! Many thanks for your prompt response

**Reply by threehams** [reply]
I'm impatient. How can I help with the TypeScript part of v5 now?

**Reply by davidkpiano** [reply]
> I'm impatient. How can I help with the TypeScript part of v5 now?

Feel free to jump into the `next` branch and make PRs there!

**Comment by bard** [comment]
@iacoware this isn't exactly what you want but still a step a up from a cast: https://github.com/davidkpiano/xstate/discussions/2071

**Reply by iacoware** [reply]
Thanks for the pointer! Type narrowing could definitely be a solution and, contrary to downcasts, you get a useful error at runtime
